api/src/main/java/org/pmiops/workbench/profile/ProfileService.java

```diff
@@ -337,4 +391,303 @@ private void validateStringLength(String field, String fieldName, int max, int m
+    if (userAuthentication.getUserType() == UserType.SERVICE_ACCOUNT) {
+      // Service accounts don't need further initialization.
+      return dbUser;
+    }
+
+    // On first sign-in, create a FC user, billing project, and set the first sign in time.
+    if (dbUser.getFirstSignInTime() == null) {
+      // If the user is already registered, their profile will get updated.
+      fireCloudService.registerUser(
+          dbUser.getContactEmail(), dbUser.getGivenName(), dbUser.getFamilyName());
+
+      dbUser.setFirstSignInTime(new Timestamp(clock.instant().toEpochMilli()));
+      // If the user is logged in, then we know that they have followed the account creation
+      // instructions sent to
+      // their initial contact email address.
+      dbUser.setEmailVerificationStatusEnum(EmailVerificationStatus.SUBSCRIBED);
+      return saveUserWithConflictHandling(dbUser);
+    }
+
+    return dbUser;
+  }
+
+  private DbUser saveUserWithConflictHandling(DbUser dbUser) {
+    try {
+      return userDao.save(dbUser);
+    } catch (ObjectOptimisticLockingFailureException e) {
+      log.log(Level.WARNING, "version conflict for user update", e);
+      throw new ConflictException("Failed due to concurrent modification");
+    }
+  }
+
+  public DbUser requestBetaAccess() {
+    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+    DbUser user = userProvider.get();
+    if (user.getBetaAccessRequestTime() == null) {
+      log.log(
+          Level.INFO,
+          String.format("Sending beta access request email to %s.", user.getContactEmail()));
+      try {
+        mailServiceProvider.get().sendBetaAccessRequestEmail(user.getUsername());
+      } catch (MessagingException e) {
+        throw new EmailException("Error submitting beta access request", e);
+      }
+      user.setBetaAccessRequestTime(now);
+      user = saveUserWithConflictHandling(user);
+    }
+    return user;
+  }
+
+  public DbUser submitDataUseAgreement(Integer dataUseAgreementSignedVersion, String initials) {
+    DbUser user =
+        userService.submitDataUseAgreement(
+            userProvider.get(), dataUseAgreementSignedVersion, initials);
+    return saveUserWithConflictHandling(user);
+  }
+
+  private void checkUserCreationNonce(DbUser user, String nonce) {
+    if (Strings.isNullOrEmpty(nonce)) {
+      throw new BadRequestException("missing required creationNonce");
+    }
+    if (user.getCreationNonce() == null || !nonce.equals(user.getCreationNonce().toString())) {
+      throw new UnauthorizedException("invalid creationNonce provided");
+    }
+  }
+
+  private boolean userHasEverLoggedIn(
+      com.google.api.services.directory.model.User googleUser, DbUser user) {
+    return user.getFirstSignInTime() != null || !googleUser.getChangePasswordAtNextLogin();
+  }
+
+  // TODO: should not be aware of ResponseEntity at the Service level
+  private ResponseEntity<Void> resetPasswordAndSendWelcomeEmail(String username, DbUser user) {
```

Comment:
We should fix these in a future PR

Reply:
I'd prefer you don't introduce ResponseEntity into the service level. It's a simple operation to return the inner value here and wrap it in a ResponseEntity at the top level.